torch/optim/swa_utils.py

```diff
@@ -125,14 +149,26 @@ def update_parameters(self, model):
             itertools.chain(model.parameters(), model.buffers())
             if self.use_buffers else model.parameters()
         )
+        self_param_detached = []
+        model_param_detached = []
         for p_swa, p_model in zip(self_param, model_param):
             device = p_swa.device
             p_model_ = p_model.detach().to(device)
+            model_param_detached.append(p_model_)
+            self_param_detached.append(p_swa.detach())
             if self.n_averaged == 0:
                 p_swa.detach().copy_(p_model_)
-            else:
-                p_swa.detach().copy_(self.avg_fn(p_swa.detach(), p_model_,
+
+        if self.n_averaged > 0 and self.avg_fn is not None:
+            for p_swa, p_model in zip(self_param_detached, model_param_detached):
+                p_swa.detach().copy_(self.avg_fn(p_swa.detach(), p_model,
                                                  self.n_averaged.to(device)))
+        elif self.n_averaged > 0 and self.multi_avg_fn is not None:
+            grouped_self = _group_tensors_by_device_and_dtype([self_param_detached])
+            grouped_model = _group_tensors_by_device_and_dtype([model_param_detached])
+            for device, dtype in grouped_self.keys():
+                self.multi_avg_fn(grouped_self[(device, dtype)][0], grouped_model[(device, dtype)][0], self.n_averaged.to(device))
```

Comment:
```suggestion
            grouped_tensors = _group_tensors_by_device_and_dtype([self_param_detached, model_param_detached])
            for ((device, _), [self_params, model_params]) in grouped_self.items():
                self.multi_avg_fn(self_params, model_params, self.n_averaged.to(device))
```
This should be possible more better clarity